src/session.c

```diff
@@ -409,11 +409,15 @@ static CK_RV token_login(P11PROV_CTX *provctx, struct p11prov_slot *slot,
             ret = CKR_GENERAL_ERROR;
             goto done;
         }
+        if (cb_pin_len == 0) {
+            ret = CKR_USER_NOT_LOGGED_IN;
+            goto done;
+        }
 
         pin = (CK_UTF8CHAR_PTR)cb_pin;
         pinlen = cb_pin_len;
     } else {
-        ret = CKR_GENERAL_ERROR;
+        ret = CKR_USER_NOT_LOGGED_IN;
```

Comment:
my only concern is if the `CKR_USER_NOT_LOGGED_IN` is the correct error to return here (I used it mostly as a fast hack to try if it will do something in my patch set). The [PKCS#11 specs](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/errata01/os/pkcs11-base-v2.40-errata01-os-complete.html#_Toc441755792) say

> ·         `CKR_USER_NOT_LOGGED_IN`: The desired action cannot be performed because the appropriate user (or an appropriate user) is not logged in.  One example is that a session cannot be logged out unless it is logged in.  Another example is that a private object cannot be created on a token unless the session attempting to create it is logged in as the normal user.  A final example is that cryptographic operations on certain tokens cannot be performed unless the normal user is logged in.

which really does not match what we are trying to do here. But good is that the C_Login should not return this value if I see right.

Reading through other return codes, using something more "unique" would be better, for example "CKR_CANCEL"?  Or `CKR_FUNCTION_CANCELED`

> ·         `CKR_FUNCTION_CANCELED`: The function was canceled in mid-execution.  This happens to a cryptographic function if the function makes a CKN_SURRENDER application callback which returns CKR_CANCEL (see CKR_CANCEL). **It also happens to a function that performs PIN entry through a protected path.** The method used to cancel a protected path PIN entry operation is device dependent.

Or just use some `CKR_VENDOR_DEFINED + X` to avoid any conflicts altogether.

Reply:
We could define our own error space, I've been tempted a few times already, but I see no problem in reusing CKR ones for our purpose, given we are never returning these errors to an application anyway (we are the consumer, not a pkcs#11 module).

I use CKR_CANCEL in other places, but I wanted an error that very specifically is about "login", and this makes it clear what happened even though it is slightly misused.